pkg/job/scraper.go

```diff
@@ -0,0 +1,241 @@
+package job
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/job/cloudwatchrunner"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/clients/account"
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/logging"
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/model"
+)
+
+type Scraper struct {
+	jobsCfg       model.JobsConfig
+	logger        logging.Logger
+	runnerFactory runnerFactory
+}
+
+type runnerFactory interface {
+	GetAccountClient(region string, role model.Role) account.Client
+	NewResourceMetadataRunner(logger logging.Logger, region string, role model.Role) ResourceMetadataRunner
+	NewCloudWatchRunner(logger logging.Logger, region string, role model.Role, job cloudwatchrunner.Job) CloudwatchRunner
+}
+
+type ResourceMetadataRunner interface {
+	Run(ctx context.Context, region string, job model.DiscoveryJob) ([]*model.TaggedResource, error)
+}
+
+type CloudwatchRunner interface {
+	Run(ctx context.Context) ([]*model.CloudwatchData, error)
+}
+
+func NewScraper(logger logging.Logger,
+	jobsCfg model.JobsConfig,
+	runnerFactory runnerFactory,
+) *Scraper {
+	return &Scraper{
+		runnerFactory: runnerFactory,
+		logger:        logger,
+		jobsCfg:       jobsCfg,
+	}
+}
+
+func (s Scraper) Scrape(ctx context.Context) ([]model.TaggedResourceResult, []model.CloudwatchMetricResult, []Error) {
+	// Setup so we only do one GetAccount call per region + role combo when running jobs
+	roleRegionToAccount := map[model.Role]map[string]func() (string, error){}
+	jobConfigVisitor(s.jobsCfg, func(_ any, role model.Role, region string) {
+		if _, exists := roleRegionToAccount[role]; !exists {
+			roleRegionToAccount[role] = map[string]func() (string, error){}
+		}
+		roleRegionToAccount[role][region] = sync.OnceValues[string, error](func() (string, error) {
```

Comment:
(nit) I think it might happen that a func for a given role+region has already been assigned as part of the sequential visitor work, we could check if the entry already exists?